cdap-runtime-ext-dataproc/src/main/java/io/cdap/cdap/runtime/spi/provisioner/dataproc/ComputeNetworkClient.java

```diff
@@ -0,0 +1,442 @@
+      // network is a url like https://www.googleapis.com/compute/v1/projects/<project>/<region>/networks/<name>
+      // we want to get the last section of the path and compare to the configured network name
+      int idx = firewall.getNetwork().lastIndexOf('/');
+      String networkName = idx >= 0 ? firewall.getNetwork().substring(idx + 1) : firewall.getNetwork();
+      if (!networkName.equals(networkInfo.getName())) {
+        continue;
+      }
+
+      String direction = firewall.getDirection();
+      if (!"INGRESS".equals(direction) || firewall.getAllowed() == null) {
+        continue;
+      }
+
+      if (isInternalIPOnly) {
+        // If the Dataproc cluster is using internal IP only, we are only interested in firewall rule that has source
+        // IP range overlap with one of the private IP block or doesn't have source IP at all.
+        // This is because if Dataproc cluster is using internal IP, the CDAP itself must be running inside one of the
+        // private IP blocks in order to be able to communicate with Dataproc.
+        try {
+          List<IPRange> sourceRanges = Optional.ofNullable(firewall.getSourceRanges())
+            .map(DataprocUtils::parseIPRanges)
+            .orElse(Collections.emptyList());
+
+          if (!sourceRanges.isEmpty()) {
+            boolean isPrivate = PRIVATE_IP_RANGES.stream()
+              .anyMatch(privateRange -> sourceRanges.stream().anyMatch(privateRange::isOverlap));
+            if (!isPrivate) {
+              continue;
+            }
+          }
+        } catch (Exception e) {
+          LOG.warn("Failed to parse source ranges from firewall rule {}", firewall.getName(), e);
+        }
+      }
+
+      for (Firewall.Allowed allowed : firewall.getAllowed()) {
+        String protocol = allowed.getIPProtocol();
+        boolean addTag = false;
+        if ("all".equalsIgnoreCase(protocol)) {
+          requiredPorts.clear();
+          addTag = true;
+        } else if ("tcp".equalsIgnoreCase(protocol) && isPortAllowed(allowed.getPorts(), FirewallPort.SSH.port)) {
+          requiredPorts.remove(FirewallPort.SSH);
+          addTag = true;
+        }
+        if (addTag && firewall.getTargetTags() != null && !firewall.getTargetTags().isEmpty()) {
+          tags.add(firewall.getTargetTags().iterator().next());
+        }
+      }
+    }
+
+    if (!requiredPorts.isEmpty()) {
+      String portList = requiredPorts.stream().map(p -> String.valueOf(p.port)).collect(Collectors.joining(","));
+      throw new IllegalArgumentException(String.format(
+        "Could not find an ingress firewall rule for network '%s' in project '%s' for ports '%s'. " +
+          "Please create a rule to allow incoming traffic on those ports for your IP range.",
+        networkInfo.getName(), conf.getNetworkHostProjectID(), portList));
+    }
+    return tags;
+  }
+
+  @Override
+  public boolean isNetworkPeered(String systemProjectId, String systemNetwork) {
```

Comment:
this is one of the few parts that is not cut and paste from DataprocClient. This is taken from the end of DataprocClient.isInternalIPOnly()